econml/iv/dr/_dr.py

```diff
@@ -73,83 +246,1246 @@ def _effect_estimate(self, nuisances):
+        clipping of the covariate for regions with low "overlap", to reduce variance
+
+    opt_reweighted : bool, optional, default False
+        Whether to reweight the samples to minimize variance. If True then
+        model_final.fit must accept sample_weight as a kw argument. If True then
+        assumes the model_final is flexible enough to fit the true CATE model. Otherwise,
+        it method will return a biased projection to the model_final space, biased
+        to give more weight on parts of the feature space where the instrument is strong.
+
+    discrete_instrument: bool, optional, default False
+        Whether the instrument values should be treated as categorical, rather than continuous, quantities
+
+    discrete_treatment: bool, optional, default False
+        Whether the treatment values should be treated as categorical, rather than continuous, quantities
+
+    categories: 'auto' or list, default 'auto'
+        The categories to use when encoding discrete treatments (or 'auto' to use the unique sorted values).
+        The first category will be treated as the control treatment.
+
+    cv: int, cross-validation generator or an iterable, optional, default 2
+        Determines the cross-validation splitting strategy.
+        Possible inputs for cv are:
+
+        - None, to use the default 3-fold cross-validation,
+        - integer, to specify the number of folds.
+        - :term:`CV splitter`
+        - An iterable yielding (train, test) splits as arrays of indices.
+
+        For integer/None inputs, if the treatment is discrete
+        :class:`~sklearn.model_selection.StratifiedKFold` is used, else,
+        :class:`~sklearn.model_selection.KFold` is used
+        (with a random shuffle in either case).
+
+        Unless an iterable is used, we call `split(concat[W, X], T)` to generate the splits. If all
+        W, X are None, then we call `split(ones((T.shape[0], 1)), T)`.
+
+    mc_iters: int, optional (default=None)
+        The number of times to rerun the first stage models to reduce the variance of the nuisances.
+
+    mc_agg: {'mean', 'median'}, optional (default='mean')
+        How to aggregate the nuisance value for each sample across the `mc_iters` monte carlo iterations of
+        cross-fitting.
+
+    random_state: int, :class:`~numpy.random.mtrand.RandomState` instance or None, optional (default=None)
+        If int, random_state is the seed used by the random number generator;
+        If :class:`~numpy.random.mtrand.RandomState` instance, random_state is the random number generator;
+        If None, the random number generator is the :class:`~numpy.random.mtrand.RandomState` instance used
+        by :mod:`np.random<numpy.random>`.
+    """
+
+    def __init__(self, *,
+                 model_y_xw="auto",
+                 model_t_xw="auto",
+                 model_z_xw="auto",
+                 model_t_xwz="auto",
+                 model_tz_xw="auto",
+                 prel_model_effect,
```

Comment:
We should do the exact same thing here as what we did in the IntentToTreatDRIV:
if `prel_model_effect` is auto, then we should instantiate it with a DRIV class, with a DummyCate prel_model_effect and with the same options as the options passed by the user. 

Actually an even better approach: ask them to pass a `flexible_model_final` as we do in the IntentToTreatDRIV

Also ask them for an option for prel_model_effect_type = {'dmliv', 'driv'}

If prel_model_effect_type = 'dmliv', then we use a dmliv as a first stage estimator of the effect. In this case, we also need them to provide a model_t_xwz, whether projection is true or false, since that is needed by dmliv. And in dmliv, we use the flexible_model_final as the prelim estimator model_final.

If prel_model_effect_type='driv', then we use driv as the first stage estimator with opt_reweighted=True and with exactly the same remaining options as passed here.

Reply:
Otherwise currently it's too much to ask from the user to provide a prel_model_effect.

Reply:
So a proposed API:
```python

class _DRIV:
    same as current DRIV

class DRIV:
    def __init__(self, *,
                 model_y_xw="auto",
                 model_t_xw="auto",
                 model_z_xw="auto",
                 model_t_xwz="auto",
                 model_tz_xw="auto",
                 prel_cate_estimator='driv', # one of {'driv', 'dmliv'}
                 prel_cv=1, # cv in the preliminary estimator
                 prel_opt_reweighted=True, # whether to opt reweight prel DRIV
                 flexible_model_effect, # Any estimator that can accept sample weights
                 model_final,
                 projection=False,
                 featurizer=None,
                 fit_cate_intercept=False,
                 cov_clip=0.1,
                 opt_reweighted=False,
                 discrete_instrument=False,
                 discrete_treatment=False,
                 categories='auto',
                 cv=2,
                 mc_iters=None,
                 mc_agg='mean',
                 random_state=None):
```
If `prel_cate_estimator='driv'`, then we use `_DRIV` as preliminary estimator with `_DummyCATE` as `prel_model_effect` and with `flexible_model_effect` as `model_final` and `cv=prel_cv` and `opt_reweighted=prel_opt_reweighted`. The rest of the parameters of the preliminary DRIV are propagated to be the same as the final stage DRIV.

If `prel_cate_estimator='dmliv'`, then we use `NonParamDMLIV` as preliminary estimator with `flexible_model_effect` as `model_final` and `cv=prel_cv` and `model_y_xw=self.model_y_xw`,
`model_t_xw=self.model_t_xw`, `model_t_xwz=self.model_t_xwz`.

Note that if `prel_cate_estimator='dmliv'`, then we definitely need `self.model_t_xwz` irrespective of `projection=True/False`. Thus if `prel_cate_estimator='dmliv'`, we should allow the user to change `self.model_t_xwz` to something other than auto.

Reply:
Great suggestion! Done!